### PR TITLE
pe-wired-factory

### PR DESCRIPTION
**Pull request by pauleveritt**
First step on a factory decorator with __wired_factory__ support.

### CODE REVIEW

**Comment by pauleveritt** [comment]
@mmerickel I cleaned up some of the tox complaints (venusian imports, using variable annotations before 3.6, black, etc.). Though for some reason tox still complains about coverage on tests/* but when I run locally it doesn't.

I now moved the `__wired_factory__` sniffing to where it should be (`container.py`) and added a test in `text_container.py`.

I think it's ready to be looked at.

**Comment by mmerickel** [comment]
this PR looks good to me... lots of extra unrelated changes but the protocol and factory are good.

I think I'd like to rename the `factory` to `service_factory` to match the wired naming right now of `ServiceRegistry` and `ServiceContainer`. Could you make that happen before merging this?

**Comment by pauleveritt** [comment]
Want me to move the unrelated commits out?

Also, are you ok with me doing examples this way? If so, I'll write some docs as part of this PR and include the examples.

There's more I could do, like examples showing factory functions with `__wired_factory__` and/or decorator, from your second pastebin.

**Comment by mmerickel** [comment]
I think this strategy for docs/examples is good, feel free to improve them and let me know when it's ready.

File: src/wired/decorators.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        self.name = name
+
+    def __call__(self, wrapped):
+        def callback(scanner: venusian.Scanner, name: str, cls):
+            registry = getattr(scanner, 'registry')
+            # If there is a for_ use it, otherwise, register for the same
+            # class as the instance
+            for_ = self.for_ if self.for_ else cls
+
+            _factory = getattr(cls, '__wired_factory__', cls)
```
Comment (mmerickel) [inline src/wired/decorators.py:34]: this line is not necessary since `register_factory` will already detect it

**Pull request merged**